Defines

ciechowoj opened the pull request:
Implement translation of `functional` macrodefinitions.

jacob-carlborg commented on clang/Token.d:
```diff
@@ -18,6 +18,20 @@ import clang.Util;
+
+TokenKind toD(CXTokenKind kind)
+{
+    return cast (TokenKind) kind;
```
Please don't put a space after `cast`.

ciechowoj replied:
Fixed it in separate commit. I'll rebase it after the review is complete.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1404 @@
+    return false;
+}
+
+bool accept(ref TokenRange tokens, ref string spelling, TokenKind kind)
```
Doesn't this cause a conflict with the first overload of `accept`?

ciechowoj replied:
Doesn't seem so, however I hope it isn't due to some bug in dmd. The first overload requires explicit template parameters. I can change it, if you want.

jacob-carlborg replied:
If you don't specify any template parameters (since it's a variadic template parameter) it should cause a conflict. If the intention is that at least one template parameter should always be passed, then add a template constraint to enforce that.

ciechowoj replied:
I'm not sure if [this](https://dlang.org/spec/template.html#variadic_template_specialization) does'n cover the case, if we consider a non-template version as a template without any parameters.

ciechowoj replied:
Anyway, I fixed it.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1404 @@
+
+    string spelling;
+
+    override string transl(ref bool[string] imports)
```
Please avoid abbreviations.

ciechowoj replied:
Is `Expr` suffix on some types ok?

jacob-carlborg replied:
Yeah, I guess that's ok.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,475 @@
+}
+D", q"C
+#define FOO() 0
+C");
```
I don't see a difference compared to the previous assertion.

ciechowoj replied:
Fixed.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,475 @@
+C");
+
+    assertTMD(q"D
+int FOO(int a)
```
How do you determine that `a` should be `int`. Based on `1` in the body? What if a long is passed when calling the macro, that will not compile in D. Isn't using a template closer to how the macro works?

ciechowoj replied:
You are right, I probably wanted be too smart.

ciechowoj replied:
Fixed.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,475 @@
+    assertTMD(q"D
+string STRINGIZE(T)(in T major, in T minor)
+{
+    import std.conv;
```
When doing a local import, it's recommended to do a selective import as well.

ciechowoj replied:
Fixed.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,475 @@
+alias ulong foo_t;
+
+enum UINT_31_MAX = cast(uint_32) 0x7fffffffL;
+enum UINT_32_MAX = cast(uint_32) (-1);
```
Is the parentheses around `-1` necessary here? You seem to have been able remove redundant parentheses in other cases.

ciechowoj replied:
Fixed.

jacob-carlborg commented:
Shouldn't all macros take their parameters by reference?

ciechowoj commented:
> Shouldn't all macros take their parameters by reference?

Macros are mostly pure functions (at least the currently handled ones), I don't see any advantage (maybe speed) and you cannot pass rvalue to reference parameter.

jacob-carlborg commented:
> Macros are mostly pure functions (at least the currently handled ones), I don't see any advantage (maybe speed) and you cannot pass rvalue to reference parameter.

You have a test for a macro `#define foo(x) ++x` which is translated to `auto foo(T)(in T x){ return ++x; }`. That has not the same behaviour as the C macro. First problem, the parameter is declared as `in` which means `scope const`. The body won't compile since you cannot modify a const parameter. Second problem is when the macro/function is used:

```d
int a = 0;
foo(a);
assert(a == 1);
```

The above assertion will pass in C since the macro is expanded inline. But it will fail in D since the integer is copied to the function and the caller won't see the modification.

ciechowoj commented:
Good point. I'll see what can be done about it, but it do not seem to be trivial to fix : /.

jacob-carlborg commented:
I think the correct translation is to take all arguments by reference, unless you can figure out that the body doesn't modify the arguments. Using `ref`instead of `in` should be a trivial fix. Figuring out exactly when `ref` is needed, not so much 😃. For me it's perfectly acceptable to use `ref` for all parameters. If you want to and have the time, you can come back later and try to only insert `ref` when is actually necessary.

ciechowoj commented:
Ok, I may change `in` to `ref` and fix it later in another PR. What do you think?

ciechowoj commented:
Ok, I fixed if for `++` and `--`, currently other assigning operators aren't supported, so the risk that reference is necessary and it isn't translated that way is small. I corrected the naming of template parameters too (they have different names now). 

To illustrate problem with references consider following code:

```
#define foo(x) ((x) + 1)
#define bar(x) ((x) * 2)
#define baz(x) (foo(bar(x)) % 3)
```

If all parameters are by reference straightforward translation of `baz` will not work.

```
auto baz(T)(ref T x)
{
    return foo(bar(x)) % 3;
}
```

RValue returned by `bar` cannot be passed as reference parameter of `foo`. D doesn't allow temporaries to be passed to reference parameters.

jacob-carlborg commented:
> RValue returned by bar cannot be passed as reference parameter for foo. D doesn't allow temporaries to be passed to reference parameters.

What about [auto ref](https://dlang.org/spec/template.html#auto-ref-parameters)?

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+            return format("(%s)", subexpr.translate(imports));
+    }
+
+    bool surplus() const
```
What does `surplus` mean, is it an abbreviation of something?

ciechowoj replied:
Synonym to `unused`/`excessive`.

[1](http://www.merriam-webster.com/dictionary/surplus)
[2](http://www.thesaurus.com/browse/surplus)

jacob-carlborg replied:
Hmm, never heard of it before.

ciechowoj commented:
Looks promising, I'll read more about it and  change the code tomorrow.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+    void guessParamTypes(ref ExprType[string] params, ExprType type)
+    { }
+
+    override string toString() { return ""; }
```
Please put the body on its own line, fully expanded as the rest of the methods.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+    import std.format : format;
+    import std.array : join;
+
+    return map!(a => format("%s %s", types[a].asParamType(), a))(params).join(", ");
```
I prefer to use UFCS for algorithm functions.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+    import std.algorithm.searching : count;
+    import std.format : format;
+    import std.array : join;
+    import std.array : appender;
```
You can put `appender` on the previous line.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+}
+
+
+
```
Please remove excessive newlines.

ciechowoj replied:
Fixed.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,457 @@
+    assertTMD(q"D
+D", q"C
+#define FOO
+C");
```
Isn't it more natural to have the C code as the first parameter? Like here https://github.com/jacob-carlborg/dstep/blob/master/unit_tests/UnitTests.d#L14-L16.

ciechowoj replied:
I would like to change all the assertion to look as here. I believe the order in `assertTranslates` was unlucky. Most unittesting frameworks takes *expected* value as first parameter.

jacob-carlborg replied:
That's true. But I think in this case, when there are two string literals, the input and the expected, it feels more natural with the input value (the C code) first. The actual result is not in the assertion, it's the input value and the expected result. So it's still not like other frameworks.

ciechowoj replied:
You have a point. Fixed.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,457 @@
+    assertTMD(q"D
+size_t FOO(T)(in T a)
+{
+    return sizeof(a);
```
`sizeof` in D is a property, i.e. `a.sizeof`.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -204,3 +225,58 @@ bool isGlobal(Cursor cursor)
+    }
+}
+
+bool[string] collectTypeNames(TranslationUnit translUnit)
```
Could you please document/describe how these "collect" functions work.

ciechowoj replied:
Done.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1403 @@
+import dstep.translator.Context;
+import dstep.translator.Output;
+
+bool accept(Args...)(ref TokenRange tokens, ref string spelling, TokenKind kind) if (Args.length > 0)
```
Could you please document/describe how these "accept" functions work.

ciechowoj replied:
Ok, I'll provide documentation tomorrow.

ciechowoj replied:
Done.

ciechowoj commented:
> What about auto ref?

Changed all arguments to `auto ref`.

ciechowoj commented:
Unrelated stuff:

I tried to generate documentation with `dub build --build=docs` and it doesn't work (dub says that dmd returns 1 in some point). Have you ever tried to generate docs for dstep? I suppose it would be nice if it works.

jacob-carlborg commented:
No, I have not tried generating the documentation. You can try with the verbose flag to get the actual commands Dub is running. Then try manually running the individual commands to see if you get some more information.

ciechowoj commented:
I've just tried to do like so, but it seems like problem with dmd and beyond printing some warnings from dependencies it doesn't say nothing useful : P.

jacob-carlborg commented:
I get some errors if I remove the warnings flag (-w), mostly from Tango 😃.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1429 @@
+  * The accept family of functions parse tokens. In the case of successful parsing
+  * advances beginning of tokens to the next token and returns true.
+  *
+  * The parsing is successful, if the first token in tokens has a kind kind and
```
* "kind" is repeated twice
* "has a kind" is that any kind of the given kind

ciechowoj replied:
I wanted to say a kind `kind`. Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1429 @@
+  * advances beginning of tokens to the next token and returns true.
+  *
+  * The parsing is successful, if the first token in tokens has a kind kind and
+  * its spelling matches on of the strings passed as Args.
```
on -> one ?

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1429 @@
+  * It assigns the spelling of the token to the spelling parameter.
+  */
+
+bool accept(Args...)(ref TokenRange tokens, ref string spelling, TokenKind kind) if (Args.length > 0)
```
Is there a reason when `Args` is a template parameter instead of `string[]...`?

ciechowoj replied:
The only reason is performance (some branching can be avoided), if you consider it premature optimization, I can change it.

jacob-carlborg replied:
Most of my concern is that the API of the function is not very clear. There's no indication in the documentation or in a template constraint that `Args` should only be strings. The name `Args` is also very generic, i.e. that you could pass in any type of arguments.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1429 @@
+import dstep.translator.Output;
+
+/**
+  * The accept family of functions parse tokens. In the case of successful parsing
```
Please align the `*` to the first `*` in `/**`. This will match the module header/documentation.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -204,3 +225,66 @@ bool isGlobal(Cursor cursor)
+  *
+  * The type names are required for parsing C code (e.g. macro definition bodies),
+  * as C grammar isn't context free.
+  */
```
* Same here with the alignment of `*`
* I still don't understand how this function is used

ciechowoj replied:
To parse C-style casts (and `sizeof`s) you need to have information about type names. Without it you do not know if `(int)(4)` is a cast or a function call. This function collects that info.

jacob-carlborg replied:
So it's basically a list of all the types in the translation unit? And it's an associative array to emulate a set?

ciechowoj replied:
Exactly.

jacob-carlborg replied:
Ok, now I get it 😃.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1429 @@
+ * The accept family of functions parse tokens. In the case of successful parsing
+ * advances beginning of tokens to the next token and returns true.
+ *
+ * The parsing is successful, if the first token in tokens has is of specified kind
```
* has is -> is?
* specified -> the specified

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
+        result["unsigned long long"] = true;
+        result["float"] = true;
+        result["double"] = true;
+    }
```
* It seems that (un)signed can be specified in any position, before, after and in between (in the case of long long)
* `short int`, `long int`, `long double` are also valid types
* `const` can be specified for any type
* `__complex__` and `_Complex` in any combination of the floating point types are valid as well. The floating point type does not need to be specified
* `_Bool` and `bool` (if `<stdbool.h>` is included) are valid types in C99. I tried compiling code containing these types with `-std=c89 -Wall` and got no errors or warnings

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
+                    break;
+
+                default:
+                    break;
```
What about unions and enums?

ciechowoj replied:
Right, it is a bit leaky at the moment, there is probably a problem with tag namespace of strucutres (e.g. `struct Foo x;`). I'll fix it when I cope with constraining that `Args...` thing to be a string...

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
+ * as C grammar isn't context free.
+ */
+
+bool[string] collectTypeNames(TranslationUnit translUnit)
```
I found this snippet in the newsgroup as way to emulate sets. It clearly shows indentation that it's a set and not an associative array.

```d
alias Set(T) = void[0][T];

void add(T)(ref void[0][T] set, T value) {
    set[value] = (void[0]).init;
}

bool contains(T)(inout(void[0][T]) set, T value) {
    return (value in set) !is null;
}

Set!(string) a;
a.add("asd");
assert(a.contains("asd"));
a.remove("asd");
```

ciechowoj replied:
> It seems that (un)signed can be specified in any position, before, after and in between (in the case of long long)
> short int, long int, long double are also valid types
> const, volatile and _Atomic (C11) can be specified for types as well
> __complex__ and _Complex in any combination of the floating point types are valid as well. The > floating point type does not need to be specified
> _Bool and bool (if <stdbool.h> is included) are valid types in C99. I tried compiling code containing these types with -std=c89 -Wall and got no errors or warnings
> When a type consist of multiple words, i.e. long int, any order of the words are valid
> What about the spaces. If the source code specifies "long long" (seems like multiple spaces are removed by GitHub), will the spelling contain that as well or does Clang remove the extra spaces?

> What about unions and enums?

Well, I need to write parsing of the types. It will not be hard, but there is another problem. Even if I parse the the types, then I need to translate them. But, there are routines for translating types in dstep already, most in `dstep/translator/Type.d`. But they only accept information in the form of `Type` and `Cursor` objects. It would be stupid to have separate code path to translate types in defines.

This is related to what I said in #65 (desing problem). One way would be to make all the functions templates, but it doesn't seem doable in practice.

I propose to change a `Type` type from struct to class and then create two new classes that inherits from `Type`: `CLangType` and let say `UserType`. `CLangType` would talk to cx and access clang interface directly, `UserType` would have a constructor allowing an user to create it and fill with data directly.

Another way would be to create a `TypeImpl` struct and store both pointer to `TypeImpl` and `cx` inside `Type` and in every method check which one to use with if.

```
struct TypeImpl
{
    string spelling;
    bool isConst;
    etc...
};

struct Type
{
    mixin CX;
    TypeImpl* impl;

    string spelling() 
    {
        return impl ? impl.spelling : /* get spelling from cx */;
    }
}

```

Another way is to fill `Type` with all required fields and copy them on creation.

```
struct Type
{
    string spelling;
    bool isConst;
    bool isVolatile;
    Type* canonicalType;
    etc.

    Type(void* returned_from_clang)
    {
        // get data from libclang and fill fields
    }
}
```

What do you think about it?

jacob-carlborg replied:
I don't like the second suggestion with `TypeImpl`. I'm leaning towards the first suggestion with classes. The last suggestion isn't so bad either. The last one has the advantage that we don't need to come up with names for two new classes (I'm not a big fan of `UserType`).

ciechowoj commented:
I would prefer the first one to, its easiest to implement. The name `UserType` can be changed if you have better idea, I have mixed feeling about it too.

ciechowoj commented:
Maybe `CustomType` or `ForeignType` (as type outside clang).

jacob-carlborg commented:
I think `CustomType`, and the casing for the other one should be `ClangType`.

ciechowoj commented:
Ok.

ciechowoj commented:
Eventually I used third suggestion..., I started with changing the struct to the class but it doesn't wont to work so well, there is too much difference between reference and value semantics.

cast operator (in macro-definitions) works only for basic types for now. Structs, unions enums and pointers are on their way.

ciechowoj commented:
Ok, I added support for struct, unions, enums and typedefs.

ciechowoj commented:
> const, volatile and _Atomic (C11) can be specified for types as well
volatile isn't supported by D, there is no API for _Atomic in libclang so I'm giving up on them.

jacob-carlborg commented:
> volatile isn't supported by D, there is no API for _Atomic in libclang so I'm giving up on them.

The question is if it needs to be recognized as part of the type anyway. I'm not sure if `const` can be kept in the translation, because const is transitive in D, while not in C. In the same sense it might be necessary to drop `volatile` and `_Atomic` as well i the translation.

ciechowoj commented:
Yes, situation with volatile and _Atomic is pretty clear, but I have no idea what to do about `const`...

ciechowoj commented:
I would prefer not to change anything about translation of `const`, as this PR is about parsing macro definitions and translation code is as it was before.

jacob-carlborg commented:
> I would prefer not to change anything about translation of const, as this PR is about parsing macro definitions and translation code is as it was before.

Absolutely.

jacob-carlborg commented on clang/Type.d:
```diff
@@ -93,31 +110,53 @@ struct Type
     }
 
-    @property bool isExposed ()
+    @property void canonical(Type type)
```
Is it necessary to have setters, can everything be handled in the constructor?

ciechowoj replied:
Everything can be handled in constructor, however I thought it would be better to have a setters because some fields do not require initialization every time, so there would be many (a few) unused parameters in constructors of a few different versions of constructors. But I can change it.

jacob-carlborg replied:
So far a type has been immutable, in the sense that it did not have any setters.

ciechowoj replied:
Fixed.

jacob-carlborg commented on clang/Type.d:
```diff
@@ -130,13 +169,10 @@ struct Type
+        return format("Type(kind = %s, spelling = %s, isConst = %s)", kind, spelling, isConst);
     }
 }
 
```
For the future: it looks like you have changed order of the methods, which causes unnecessary noise in the diff, making it harder to review.

ciechowoj replied:
Ok, I'll try to remember about that. I've just grouped some kinds of methods together.

jacob-carlborg replied:
That's understandable, but with the amount of code in your pull requests, every diff counts. It's better to make as few changes as possible, especially when it comes to style related changes. I know that can be hard but it's better to leave that for a completely separate pull request. It's also in your best interest because it will make my life easier and speed up the review process 😃.

jacob-carlborg commented on clang/Type.d:
```diff
@@ -4,73 +4,90 @@
  * Version: Initial created: Jan 29, 2012
  * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0)
  */
+
```
I prefer that there's no newline between the documentation and the symbol it's attached to.

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -44,8 +49,19 @@ class Context
+
+    public string macroLinkagePrefix()
+    {
+        return macroLinkage == "" ? "" : macroLinkage ~ " ";
```
Why isn't the space included from the beginning?

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/Context.d:
```diff
@@ -33,8 +34,12 @@ class Context
 
-    public this(TranslationUnit translUnit, Options options, Translator translator)
+    public MacroDefinition[string] macroDefinitions;
+    string macroLinkage = "extern (D)";
```
Is it necessary to set the linkage?

ciechowoj replied:
The whole file is wrapped to `extern (C):`, so I thought it would be wise to override it for the translations of macros to not cause a clash with other symbols (e.g. from static library, mangled in C-style).

> Why isn't the space included from the beginning?

I thought about a CLI switch to control that behavior and I do not want the user to be obliged to remember about formatting issues.

jacob-carlborg replied:
Do you mean that the user could set the linkage for a translated macro?

ciechowoj replied:
Yes that was the aim, I kinda forgot to add a flag to the driver code. What do you think about that idea?

jacob-carlborg replied:
Currently I cannot come up with a use case for that. Perhaps keep it simple and not make that possible.

ciechowoj replied:
Done.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -1188,7 +1221,417 @@ Expression parseCondExpr(ref TokenRange tokens, bool[string] table)
+            tokens = local;
+            return true;
+        }
+    }
```
These two if statements are exactly the same, except for the keyword `struct/union`. If you add an additional variable, `keywordType` or similar and pass that in to the `accept` function and two template arguments `struct` and `union` then `keywordType` will contain which keyword that actually match on return?

ciechowoj replied:
You are right, fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -1188,7 +1221,417 @@ Expression parseCondExpr(ref TokenRange tokens, bool[string] table)
+    if (accept!(qualifiers)(tokens, spelling))
+        return true;
+    else
+        return false;
```
Doesn't `accept` already return a bool?

ciechowoj replied:
Fixed.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -1188,7 +1221,417 @@ Expression parseCondExpr(ref TokenRange tokens, bool[string] table)
+    }
+
+    return false;
+}
```
Does this handle `long int`?

ciechowoj replied:
Yes here are tests:
[link](https://github.com/jacob-carlborg/dstep/pull/67/files#diff-21c4f0499bb291726704dbb88c44efaeR97)

jacob-carlborg replied:
Ah, I see now, thanks.

ciechowoj replied:
Eh, that link isn't very good, the tests are in unit_tests/MacroParsingTests.d:128 .

jacob-carlborg replied:
I found the test anyway 😃.

jacob-carlborg commented on unit_tests/Assert.d:
```diff
@@ -0,0 +1,143 @@
+
+    auto definition = parsePartialMacroDefinition(cursor.tokens, context.typeNames);
+
+    bool[string] imports;
```
I guess this could use the new set "type".

ciechowoj replied:
I guess you are right, fixed.

jacob-carlborg commented on clang/Type.d:
```diff
@@ -93,31 +115,39 @@ struct Type
+
+    @property Type canonical()
+    {
+        return Type(clang_getCanonicalType(cx));
```
This will only work if the type is create from a Clang type. Perhaps adding an assertion?

jacob-carlborg commented on clang/Type.d:
```diff
@@ -93,31 +115,39 @@ struct Type
     {
-        auto r = clang_getTypeDeclaration(cx);
-        return Cursor(r);
+        return Cursor(clang_getTypeDeclaration(cx));
```
Same here, will only work for Clang types.

ciechowoj replied:
I assumed that libclang detect that the type is invalid and return invalid/empty cursor/type, but now its explicit.

jacob-carlborg commented on clang/Type.d:
```diff
@@ -93,31 +117,55 @@ struct Type
     {
-        auto r = clang_getTypeDeclaration(cx);
-        return Cursor(r);
+        if (isClang)
```
Isn't it better to us an `assert`, because you're not really supposed to call this method unless the type is coming from Clang?

ciechowoj replied:
Some code other depends on this. E.g. to check if the type has associated declaration.

jacob-carlborg replied:
Ok, I see.

jacob-carlborg commented on unit_tests/MacroTranslTests.d:
```diff
@@ -0,0 +1,566 @@
+
+}
+
+// Translate cast operator (FIXME: type inference)
```
What's up with this `FIXME`?

ciechowoj replied:
It is about trying to guess the return type of macro. But I suppose it is unimportant now, as D's `auto` is good enough. I removed the fixme.

jacob-carlborg replied:
👍

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1859 @@
+            return expr;
+        }
+
+        // FIXME: unary-expression ::= sizeof ( type-name )
```
What's up with this `FIXME`?

ciechowoj replied:
Parsing of C-style sizeof is type dependent too (you need to know the names of types). I added that FIXME when translation of types was not fully implemented. But as it is handled quite good now, I've just added a commit that implements parsing and translation of type-dependent `sizeof` operators.

jacob-carlborg replied:
Is the FIXME still needed?

ciechowoj replied:
Of course not. Forgot to push.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -0,0 +1,1859 @@
+        "signed",
+        "unsigned",
+        // "__complex__", TBD
+        // "_Complex", TBD
```
What's the status of these two "complex", do you plan to implement them in this PR?

jacob-carlborg replied:
Fair enough.

jacob-carlborg commented:
👌

jacob-carlborg commented:
Could you please clean up the commits a bit and then I can merge this.

ciechowoj commented:
done

jacob-carlborg commented:
Please cleanup the commit message for the second commit.

ciechowoj commented:
Actually I intentionally left it that way, but ok : )

jacob-carlborg commented:
Ideally, the fixes you've done according to my comments should not be visible in the history.

ciechowoj commented:
Done, It has some sense.

ciechowoj commented:
Thanks.

jacob-carlborg commented:
Thanks for the great work 👍.

jacob-carlborg commented:
Well, I haven't really planned anything 😃. I could at least create a tag but building binaries is kind of a pain in the ass.

jacob-carlborg commented:
@ciechowoj could you please update the changelog as well, I always forget that. I suspecting it's not completely up to date, could you verify that?

ciechowoj commented:
I've just added a PR.
